### PR TITLE
HADOOP-17880. Build 2.10.x with docker

### PR DESCRIPTION
**Pull request by ZhendongBai**
### Description of PR
this pr support build hadoop 2.10.x with docker.

### How was this patch tested?
test on mac x86_64

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id? HADOOP-17880
- [x] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by goiri** [comment]
@GauthamBanasandra can you take a look at this PR?

**Comment by GauthamBanasandra** [comment]
@ZhendongBai could you please mention the HADOOP JIRA in the title of this PR?

File: dev-support/docker/Dockerfile [hunk 1]
```diff
@@ -18,234 +17,80 @@
 # Dockerfile for installing the necessary dependencies for building Hadoop.
 # See BUILDING.txt.
 
-FROM ubuntu:xenial
+FROM centos:7
```
Comment (GauthamBanasandra) [inline dev-support/docker/Dockerfile]: We've been using Ubuntu as the default build environment. May I know why you would want to use Centos 7 instead of Ubuntu Xenial?

**Comment by ZhendongBai** [comment]
> @ZhendongBai could you please mention the HADOOP JIRA in the title of this PR?

@GauthamBanasandra thanks a lot for your review, and just now I create a Haooop Jira issue with description, and mention it in PR description. please review again.

**Comment by GauthamBanasandra** [comment]
> > @ZhendongBai could you please mention the HADOOP JIRA in the title of this PR?
> 
> @GauthamBanasandra thanks a lot for your review, and just now I create a Haooop Jira issue with description, and mention it in PR description. please review again.

@ZhendongBai you would need to mention the Hadoop JIRA in the title of the PR, as mentioned in the description.

File: dev-support/docker/Dockerfile (hunk 1 above)
Comment (ZhendongBai) [reply dev-support/docker/Dockerfile]: @GauthamBanasandra  for ubuntu, openjdk-7 not found installation candidate, the error shows as bellow:
```
.....
ERROR [ 8/27] RUN apt-get -q update     && apt-get -q install -y --no-install-recommends openjdk-7-jdk     && apt-get clean     && rm -rf /var/lib/apt/lists/*
#11 94.95 Get:20 http://archive.ubuntu.com/ubuntu xenial-backports/universe amd64 Packages [12.7 kB]
#11 95.07 Fetched 19.4 MB in 1min 34s (205 kB/s)
#11 95.07 Reading package lists...
#11 95.99 Reading package lists...
#11 96.88 Building dependency tree...
#11 97.02 Reading state information...
#11 97.04 Package openjdk-7-jdk is not available, but is referred to by another package.
#11 97.04 This may mean that the package is missing, has been obsoleted, or
#11 97.04 is only available from another source
#11 97.04
#11 97.04 E: Package 'openjdk-7-jdk' has no installation candidate
```

**Comment by ZhendongBai** [comment]
> > > @ZhendongBai could you please mention the HADOOP JIRA in the title of this PR?
> > 
> > 
> > @GauthamBanasandra thanks a lot for your review, and just now I create a Haooop Jira issue with description, and mention it in PR description. please review again.
> 
> @ZhendongBai you would need to mention the Hadoop JIRA in the title of the PR, as mentioned in the description.

@GauthamBanasandra  ok, add the jira at the start of PR title.

File: dev-support/docker/Dockerfile (hunk 1 above)
Comment (GauthamBanasandra) [reply dev-support/docker/Dockerfile]: That's because Java 7 isn't there in the Ubuntu Xenial toolchain. Have you tried installing the Java 7 by downloading its tarball? Maybe this'll help - https://linuxconfig.org/oracle-java-jdk-7-on-ubuntu-linux-source-or-rpm-installation

I would advise against changing the base image to Centos 7 if this is the only reason. Here are my suggestions -
1. If you still want to use Centos 7, I would suggest that you create a separate Dockerfile for it (like Dockerfile_centos_7) and leave the original Dockerfile as it is.
2. Or, just install Java 7 from the tarball in Dockerfile.

Comment (ZhendongBai) [reply dev-support/docker/Dockerfile]: @GauthamBanasandra ok, thanks a lot,besides jdk7 not found problem, python pylint is not installed sucessfully,and the python dependencies have problems. So I decide to give up fixing the ubuntu bugs, and choose to  create a separate Dockerfile for centos later.

**Comment by ZhendongBai** [comment]
@GauthamBanasandra I update the PR， and add the new Dockerfile named Dockerfile_centos7, and keep the old Dockerfile unchanged, for reasons:
1. openjdk-7-jdk is not found in ppa:openjdk-r/ppa repo.
2. shellcheck  not found in ppa:jonathonf/ghc-8.0.2 repo.
3. pylint==1.9.2 cannot be installed successfully,for pip and setuptools and other unkown dependencies compatibilities.

**Comment by GauthamBanasandra** [comment]
Ok, sounds good @ZhendongBai . Since there's no CI setup for this branch. Could you please build this branch locally, `tee` the  build command and upload the build logs here?

**Comment by ZhendongBai** [comment]
@GauthamBanasandra  Ok, I rerun the docker build and mvn commands, after `./start-build-env.sh` in project root directory , logs as below: [docker_image_build.log](https://github.com/apache/hadoop/files/7149186/docker_image_build.log). After `mvn package -Pdist,native -DskipTests -Dtar -Dmaven.javadoc.skip=true` on docker container, finally, the project was built successfully, logs as below: [build.log](https://github.com/apache/hadoop/files/7149181/build.log).
 Take a look,please, and thanks a lot.

**Comment by GauthamBanasandra** [comment]
@ZhendongBai could you please use the following command to build and upload the build log?
```
mvn clean package -Dhttps.protocols=TLSv1.2 -DskipTests -Pnative,dist -Drequire.fuse -Drequire.openssl -Drequire.snappy -Drequire.valgrind -Drequire.zstd -Drequire.test.libhadoop -Pyarn-ui -Dtar > build.log 2>&1 &
```

File: dev-support/docker/Dockerfile_centos7 [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (GauthamBanasandra) [inline dev-support/docker/Dockerfile_centos7]: Could you please rename this file to `Dockerfile_centos_7`? Just so that we're consistent with the filename in trunk.

**Review by GauthamBanasandra — CHANGES_REQUESTED** [review]

File: dev-support/docker/Dockerfile_centos7 (hunk 2 above)
Comment (ZhendongBai) [reply dev-support/docker/Dockerfile_centos7]: @GauthamBanasandra I rename Dockerfile_centos7 to Dockerfile_centos_7 to keep consistent with the filename in trunk. and `mvn clean package -Dhttps.protocols=TLSv1.2 -DskipTests -Pnative,dist -Drequire.fuse -Drequire.openssl -Drequire.snappy -Drequire.valgrind -Drequire.zstd -Drequire.test.libhadoop -Pyarn-ui -Dtar -Dmaven.javadoc.skip=true > build.log 2>&1` logs here:  [build.log](https://github.com/apache/hadoop/files/7158247/build.log), and because some javadocs are illegal, and javadocs check failed, I add `-Dmaven.javadoc.skip=true` to build command. please review again, thanks.

File: BUILDING.txt [hunk 3]
```diff
@@ -21,38 +21,15 @@ The easiest way to get an environment with all the appropriate tools is by means
...
-Known issues:
-- On Mac with Boot2Docker the performance on the mounted directory is currently extremely slow.
-  This is a known problem related to boot2docker on the Mac.
-  See:
-    https://github.com/boot2docker/boot2docker/issues/593
-  This issue has been resolved as a duplicate, and they point to a new feature for utilizing NFS mounts
-  as the proposed solution:
-    https://github.com/boot2docker/boot2docker/issues/64
-  An alternative solution to this problem is to install Linux native inside a virtual machine
-  and run your IDE and Docker etc inside that VM.
```
Comment (GauthamBanasandra) [inline BUILDING.txt]: You would need to restore these lines as Hadoop 2.10.x is now buildable using `Dockerfile` on Virtualbox (as of this commit).

File: start-build-env.sh [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
```
Comment (GauthamBanasandra) [inline start-build-env.sh]: It seems like you've copied this file from trunk. Could you please explain why the changes to this file are needed?

File: dev-support/docker/Dockerfile [hunk 5]
```diff
@@ -248,4 +248,4 @@ RUN curl -L -o hugo.deb https://github.com/gohugoio/hugo/releases/download/v0.30
 COPY hadoop_env_checks.sh /root/hadoop_env_checks.sh
 RUN chmod 755 /root/hadoop_env_checks.sh
 # hadolint ignore=SC2016
-RUN echo '${HOME}/hadoop_env_checks.sh' >> /root/.bashrc
+RUN echo '${HOME}/hadoop_env_checks.sh' >> /root/.bashrc
```
Comment (GauthamBanasandra) [inline dev-support/docker/Dockerfile]: You shouldn't need to modify this file at all.

**Review by GauthamBanasandra — CHANGES_REQUESTED** [review]

File: dev-support/docker/Dockerfile (hunk 5 above)
Comment (ZhendongBai) [reply dev-support/docker/Dockerfile]: @GauthamBanasandra I will rollback the file changes.

File: BUILDING.txt (hunk 3 above)
Comment (ZhendongBai) [reply BUILDING.txt]: @GauthamBanasandra but when this PR merged, at some point, we can build hadoop 2.10.0 just with docker directly. And I don't know when the BUILDING.txt should change and how to describe that.

File: start-build-env.sh (hunk 4 above)
Comment (ZhendongBai) [reply start-build-env.sh]: @GauthamBanasandra  the line 1: see HADOOP-13963. /bin/bash is hard coded in some of the scripts. Contributed by Ajay Yadav.  and others changes can be see via git log at chunk branch. and besides this, I give up Dockerfile_aarch64, for centos 7 support arm64 platform. so I removal relational codes.

Comment (GauthamBanasandra) [reply start-build-env.sh]: `start-build-env.sh` creates the build environment based on `Dockerfile`. Since you haven't modified `Dockerfile` in this PR, you shouldn't change this file as well.

Here's my suggestion to you - let's keep the old way as it is, which is, building using Virtualbox that uses `Dockerfile`. Those who want to use Docker for building can use `Dockerfile_centos_7` directly and not through `start-build-env.sh` -
```shell
$ docker build -t hadoop-centos-7 -f .\dev-support\docker\Dockerfile_centos_7 .\dev-support\docker\
```

File: BUILDING.txt (hunk 3 above)
Comment (GauthamBanasandra) [reply BUILDING.txt]: I think we should retain the ability to build using Virtualbox and your PR would just add the ability to build using Docker, instead of deprecating the ability to build using Virtualbox. With that said, you wouldn't need to modify the docs here.

**Review by GauthamBanasandra — CHANGES_REQUESTED** [review]

File: BUILDING.txt (hunk 3 above)
Comment (ZhendongBai) [reply BUILDING.txt]: ok, I will revert the file.

**Comment by ZhendongBai** [comment]
@GauthamBanasandra pls review again, thanks a lot.

**Review by GauthamBanasandra — CHANGES_REQUESTED** [review]
Looks good to me, one final change that I would like to request from you @ZhendongBai - please change the title to something appropriate. Something like `HADOOP-17880. Build 2.10.x with docker` in both Github and JIRA. I'll approve this PR after you make this change.

**Comment by ZhendongBai** [comment]
> Looks good to me, one final change that I would like to request from you @ZhendongBai - please change the title to something appropriate. Something like `HADOOP-17880. Build 2.10.x with docker` in both Github and JIRA. I'll approve this PR after you make this change.

@GauthamBanasandra I already changed the title for PR and JIRA, because this is my first time contribute to Hadoop, cost your time too much, thanks a lot.

**Comment by ZhendongBai** [comment]
> ### Description of PR
> this pr support build hadoop 2.10.x with docker.
> 
> ### How was this patch tested?
> test on mac x86_64
> 
> ### For code changes:
> * [x]  Does the title or this PR starts with the corresponding JIRA issue id? [HADOOP-17880](https://issues.apache.org/jira/browse/HADOOP-17880)
> * [x]  Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
> * [x]  If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
> * [x]  If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

**Review by GauthamBanasandra — APPROVED** [review]

**Comment by iwasakims** [comment]
@ZhendongBai You should submit the patch based on apache:branch-2.10 instead of apache:branch-2.10.0. branch-2.10.0 is obsolete one for already released 2.10.0.

File: dev-support/docker/Dockerfile_centos_7 [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+######
+# Install packages
+######
+RUN pkg-resolver/install-maven.sh centos:7
+RUN pkg-resolver/install-cmake.sh centos:7
+RUN pkg-resolver/install-zstandard.sh centos:7
+RUN pkg-resolver/install-yasm.sh centos:7
+RUN pkg-resolver/install-protobuf.sh centos:7
+RUN pkg-resolver/install-boost.sh centos:7
```
Comment (iwasakims) [inline dev-support/docker/Dockerfile_centos_7:92]: Do we need boost for branch-2.10? If we can omit this, building time and footprint of the image can be reduced.

Comment (GauthamBanasandra) [reply dev-support/docker/Dockerfile_centos_7:92]: I upgraded Boost when I wrote this PR https://github.com/apache/hadoop/pull/2051. This was on Hadoop 3.x. Since this PR isn't backported to Hadoop 2.x, Boost isn't needed.

Comment (GauthamBanasandra) [reply dev-support/docker/Dockerfile_centos_7:92]: @ZhendongBai if you make any change to the Dockerfile_centos_7, please do a local build and upload the build log as a comment so that we can verify. We need this step since we don't have a pre-commit CI for Hadoop 2.x branch.

**Comment by ZhendongBai** [comment]
> @ZhendongBai You should submit the patch based on apache:branch-2.10 instead of apache:branch-2.10.0. branch-2.10.0 is obsolete one for already released 2.10.0.

@iwasakims thanks，I will close this pr， and commit a new pr based on branch-2.10

File: dev-support/docker/Dockerfile_centos_7 (hunk 6 above)
Comment (ZhendongBai) [reply dev-support/docker/Dockerfile_centos_7:92]: > @ZhendongBai if you make any change to the Dockerfile_centos_7, please do a local build and upload the build log as a comment so that we can verify. We need this step since we don't have a pre-commit CI for Hadoop 2.x branch.

@iwasakims @GauthamBanasandra thanks，I will try to remove boost and do a local build and upload the build log.

**Comment by ZhendongBai** [comment]
@iwasakims @GauthamBanasandra I commit a new pr base on branch-2.10, https://github.com/apache/hadoop/pull/3535, and  close this pr, please review again, thanks.
